db_sql.go

```diff
@@ -123,22 +126,71 @@ func (x *sqlUserStoreDB) ResetPasswordFinish(userId UserId, token string, passwo
 			}
 			return escan
 		}
+				
 		if everify := verifyPasswordResetToken(token, truthToken.String); everify != nil {
 			tx.Rollback()
 			return everify
 		}
+
+		if used := x.verifyPasswordArchived(userId, password); used {
+			tx.Rollback()
+			return ErrInvalidPastPassword
+		}
+
+		if earchive := x.archivePassword(tx, userId); earchive != nil {
+			tx.Rollback()
+			return earchive
+		}
+
 		if eupdate := x.setPasswordInternal(tx, userId, password); eupdate != nil {
 			tx.Rollback()
 			return eupdate
-		} else {
-			return tx.Commit()
 		}
-
+		return tx.Commit()
 	} else {
 		return etx
 	}
 }
 
+func (x *sqlUserStoreDB) archivePassword(tx *sql.Tx, userId UserId) error {
+	if _, err := tx.Exec(`INSERT INTO authpwdarchive (userid, password) SELECT userid, password FROM authuserpwd WHERE userId = $1`, userId); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (x *sqlUserStoreDB) verifyPasswordArchived(userId UserId, password string) bool {
+	archivedPasswords, err := x.getArchivedPasswordsForUser(userId)
+	if err != nil {
+		return false
+	}
+	for _, hash := range archivedPasswords {
+		if verifyAuthausHash(password, hash) {
+			return true
+		}
+	}
+	return false
+}
+
+func (x *sqlUserStoreDB) getArchivedPasswordsForUser(userId UserId) ([]string, error) {
+	passwords := make([]string, 0)
+	rows, err := x.db.Query("SELECT password FROM authpwdarchive WHERE userid = $1 ORDER BY created DESC LIMIT 15", userId)
+	if err != nil {
+		return passwords, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var passwordDbHash string
+		errScan := rows.Scan(&passwordDbHash)
```

Comment:
I see from the migration that passwordDbHash is allowed to be NULL, but this code isn't built to deal with nulls. You'll need to do use sql.NullString instead for that.
BUT.. if you actually meant to make the archived password field "NOT NULL", then obviously the migration needs to be changed.